### PR TITLE
Split TypeVisitor and use generics for context

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
**What this PR does / why we need it**:

Splits the TypeVisitor handling of `InternalTypeName` and `ExternalTypeName` into two separate code paths. Also updates TypeWalker to match.

Changes the _context_ parameter of `TypeVisitor` from `any` (or `interface{}`) to a generic type parameter, eliminating the need for runtime type casting and allowing the compiler to enforce type safety at compile time.

Other, minor, changes include

* Using `any` instead of `interface{}` 
* Removes the superfluous internal TypeVisitor flag `visitTypeNameIsIdentity` in favour of detecting a nil delegate.
* Adds helper method `VisitInternalTypeName` for visiting an InternalTypeName directly, taking care of the resulting cast.
* Changes `APIVersion` to an `InternalTypeName`

**Special notes for your reviewer**:

This came out pretty clean.

**How does this PR make you feel**:
gif

**If applicable**:
- [x] this PR contains tests

### CODE REVIEW

File: v2/tools/generator/internal/astmodel/renaming_visitor_test.go [hunk 1]
```diff
@@ -60,7 +60,7 @@ func TestRenamingVisitor_RewritesResourceOwner(t *testing.T) {
 	defs := make(TypeDefinitionSet)
 	defs.AddAll(badObject, childDef)
 
-	newName := badObject.Name().WithName("GoodName")
+	newName := badObject.Name().WithName("GoodName").(InternalTypeName)
```
Comment (matthchr) [inline v2/tools/generator/internal/astmodel/renaming_visitor_test.go:63]: Hmm, slightly unfortunate but I guess not a big deal.

File: v2/tools/generator/internal/astmodel/type_walker.go [hunk 2]
```diff
@@ -45,41 +47,43 @@ type typeWalkerState struct {
...
+		if terr != nil {
+			return nil, errors.Wrapf(terr, "visitTypeName failed for name %q", it)
+		}
 
-	it, ok := visitedTypeName.(TypeName)
-	if !ok {
-		panic(fmt.Sprintf("TypeWalker visitor visitTypeName must return a TypeName, instead returned %T", visitedTypeName))
+		itn, ok := visitedTypeName.(InternalTypeName)
+		if !ok {
+			panic(fmt.Sprintf("TypeWalker visitor visitTypeName must return a TypeName, instead returned %T", visitedTypeName))
```
Comment (matthchr) [inline v2/tools/generator/internal/astmodel/type_walker.go:83]: minor: Fix comment here? We're checking explicitly for `InternalTypeName`, so should say that?

**Review by matthchr — APPROVED** [review]

File: v2/tools/generator/internal/astmodel/renaming_visitor_test.go (hunk 1 above)
Comment (theunrepentantgeek) [reply v2/tools/generator/internal/astmodel/renaming_visitor_test.go:63]: This cast goes away in one of the later PRs in this sequence, when `WithName()` starts returning `InternalTypeName` directly.

File: v2/tools/generator/internal/astmodel/type_walker.go (hunk 2 above)
Comment (theunrepentantgeek) [reply v2/tools/generator/internal/astmodel/type_walker.go]: Done.

**Pull request merged**